### PR TITLE
[AV-1101] Interface development

### PR DESCRIPTION
**Pull request by mpkarpov-ui**
These are the changes we have made so far for the MIDAS and dummy avionics interfaces. MIDAS interface seems to be mostly functional, only having some issues seemingly from not having any serial output?

Made a small mistake -- did some datastreamer test (AV-1098) development in this branch as well.

Things to be fixed before merge:
- [x] Make sure MIDAS output is tracked properly (May need to use MIDAS hilsim branch..)
- [x] MIDAS uploading with pin emulation
- [x] Code formatting

### CODE REVIEW

File: Central-Server/hilsim-web/src/dev_config.js [hunk 1]
```diff
@@ -1,2 +1,8 @@
-export const api_url = "http://localhost"
+const server_url_request = await fetch("https://raw.githubusercontent.com/ISSUIUC/ISS_HILSIM/active_server_url/server_url.txt");
```
Comment (zyunlam) [inline Central-Server/hilsim-web/src/dev_config.js:10]: should comment this to explain why this is here

File: Test-Rack-Software/config.py [hunk 2]
```diff
@@ -2,32 +2,41 @@
...
 # Metadata needs to be imported first, so the avionics system knows what data to use for startup.
 # Import all metaconfigs
 import tars_rack.platform_meta as TARSmkIVmeta  # noqa
-
+import midas_rack.platform_meta as MIDASmkImeta  # noqa
+import dummy_rack.platform_meta as DUMMYmeta  # noqa
 
 # == EDIT THE VARIABLE BELOW TO CHANGE WHICH METADATA IS USED ==
-use_meta = TARSmkIVmeta.meta
+use_meta = MIDASmkImeta.meta
```
Comment (zyunlam) [inline Test-Rack-Software/config.py:16]: Is there a way to make it so that Midas meta and interface are combined into one variable so that we don't need to change it in 2 places?

File: Test-Rack-Software/midas_rack/interface.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+            while (time.time() < start + 10):
+                # Check for defer (This may be DRY, but there aren't many
+                # better ways to do this --MK)
+                self.server.defer()
+                if (self.server.signal_abort):
+                    self.server.state.try_transition(
+                        Datastreamer.ServerStateController.ServerState.JOB_ERROR)
+                    return False, "Abort signal recieved"
+
+                try:
```
Comment (zyunlam) [inline Test-Rack-Software/midas_rack/interface.py]: Feel like this should be it''s own function

**Review by zyunlam — CHANGES_REQUESTED** [review]

File: Test-Rack-Software/config.py (hunk 2 above)
Comment (mpkarpov-ui) [reply Test-Rack-Software/config.py:16]: Not that I can think of, it is a bit weird but combining them into one and using regular imports causes circular dependency errors. Maybe we can play around with having the interface filename in the `meta` file and dynamically `__import__`'ing it.

File: Test-Rack-Software/midas_rack/interface.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+        self.current_time += dt
+        simulation_dt = 0.01
+        # The av stack can only take a certain amount of data at a time, so we need to yield until we
+        # can safely send data.
+        if self.current_time > self.last_packet_time + simulation_dt:
+            self.last_packet_time += simulation_dt
+            if self.current_time < self.start_time + 5:
+                # Wait for 5 seconds to make sure serial is connected
+                pass
+            else:
```
Comment (mpkarpov-ui) [inline Test-Rack-Software/midas_rack/interface.py:307]: I wonder if we can drop this part of hilsim runs, now that we 1) wait for the port to be open, 2) wait for the datastreamer to be able to accept reads..

File: Central-Server/API/internal/threads.py [hunk 5]
```diff
@@ -271,10 +282,10 @@ def job_runner_loop(self):
             self.packet_buffer.clear_input_buffer()
 
             time.sleep(1)
-            if (time.time() - self.last_check > 120):
+            """if (time.time() - self.last_check > 69420): #nice
                 # Remove tars if we can't detect it
                 print("Terminated", flush=True)
-                self.terminate()
```
Comment (mpkarpov-ui) [inline Central-Server/API/internal/threads.py]: We should probably get rid of this now.

File: Central-Server/API/util/tests/datastreamer_comp.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+def run(connection: util.datastreamer_connection.DatastreamerConnection):
+    print("(tests:datastreamer_comp) Initialized", flush=True)
+    comm = connection.communicaton_channel
+    buf = pkt.DataPacketBuffer()
+
+    buf.add(pkt.SV_ACKNOWLEDGE(0))
+    buf.write_buffer_to_channel(comm)
+
+    # TODO: add an actual testing library and actual tests..
```
Comment (mpkarpov-ui) [inline Central-Server/API/util/tests/datastreamer_comp.py:14]: Future TODO, this is leaked over from a clerical error and will be addressed in a future branch

File: Test-Rack-Software/dummy_rack/interface.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        return True # Emulate successful avionics connection
+    
+    def first_setup(self) -> None:
+        time.sleep(1) # Emulate a synchronous git clone
+
+    def code_reset(self) -> None:
+        time.sleep(1) # Emulate a synchronous code reset
+
+    def power_cycle(self) -> bool:
+        time.sleep(1) # Emulate a synchronous power cycle
```
Comment (mpkarpov-ui) [inline Test-Rack-Software/dummy_rack/interface.py:35]: Should return `True`

File: Test-Rack-Software/util/remote_command.py [hunk 8]
```diff
@@ -35,9 +35,6 @@ def reset_repo():
     repo = Repo(config_meta.remote_path)
     print("(git_commands) Stashing changes..")
     repo.git.checkout(".")
-    print("(git_commands) Checking-out and pulling origin/master..")
-    repo.git.checkout("master")
-    repo.git.pull()
```
Comment (mpkarpov-ui) [inline Test-Rack-Software/util/remote_command.py:38]: Should we keep this? It lets us stay in a defined datastreamer remote state after a `reset` call

File: Test-Rack-Software/config.py (hunk 2 above)
Comment (mpkarpov-ui) [reply Test-Rack-Software/config.py:16]: I'll look into it

File: Central-Server/API/output/job_70/output.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+No data :( Why Zhu Li?
```
Comment (zyunlam) [inline Central-Server/API/output/job_70/output.txt]: Should be deleted, output directory should be ignored

File: Central-Server/hilsim-web/src/dev_config.js [hunk 10]
```diff
@@ -1,2 +1,13 @@
...
+ * Due to the nature of this project, we are interested in performing tests on seperate computers on the same network. Because of this, we
+ * spin up an http tunnel through Ngrok. Because the URL of the tunnel is random every time, we need to be able to update the target URL
+ * on both the frontend and the datastreamer. The datastreamer handles this through its own `dynamic_url.py` service, this (the frontend)
+ * handles dynamic URLs here.
+ * 
+ * To edit the dynamic url, edit `server_url.txt` on the active_server_url branch (https://github.com/ISSUIUC/ISS_HILSIM/tree/active_server_url)
+*/
+
+const server_url_request = await fetch("https://raw.githubusercontent.com/ISSUIUC/ISS_HILSIM/active_server_url/server_url.txt");
+const text = await server_url_request.text();
```
Comment (zyunlam) [inline Central-Server/hilsim-web/src/dev_config.js]: Text should be stripped because the server url might have new lines or unintended data
We should also test out the url to see if it can be connected?

File: Test-Rack-Software/midas_rack/interface.py [hunk 11]
```diff
@@ -0,0 +1,371 @@
...
+    def post_setup(self) -> None:
+        self.reset_clock()
+
+    """
+    Runs one iteration of the HILSIM loop, with a change in time of dt.
+
+    @Returns a tuple: (run_finished, run_errored, return_log)
+    """
+
+    def step(self, dt: float):
```
Comment (zyunlam) [inline Test-Rack-Software/midas_rack/interface.py:297]: Mega function is not so good

File: Test-Rack-Software/util/dynamic_url.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    @kamaji_target: Which target to retrieve the URI for? `main` will always be the main Kamaji server."""
+    if api_source is not None:
+        print("(dynamic url) Retrieving dynamic url")
+        result = requests.get(api_source)
+        try:
+            print("(dynamic url) Successfully retrieved dynamic url " + str(result.content.decode().strip()))
+            return result.content.decode().strip()
+        except Exception as e:
+            print("Failed to retrieve dynamic url:", e)
+            return api_default.strip()
```
Comment (zyunlam) [inline Test-Rack-Software/util/dynamic_url.py:17]: Remove strip

File: Test-Rack-Software/util/git_commands.py [hunk 13]
```diff
@@ -18,11 +18,11 @@ def run_script(arg_list):
     @arg_list {List[str]} -- A list of arguments to execute the command with
     @description remote_command.py is a wrapper for GitPy, which leaks memory"""
     print(
-        "(git_commands) Running script [python remote-command.py " +
+        f"(git_commands) Running script [{cfg.python_root}python remote-command.py " +
```
Comment (zyunlam) [inline Test-Rack-Software/util/git_commands.py:21]: Change this to cfg.python instead of python root, and just point to python in general

File: Test-Rack-Software/util/os_interface.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import subprocess
+
+def is_raspberrypi():
+    try:
+        with io.open('/sys/firmware/devicetree/base/model', 'r') as m:
+            if 'raspberry pi' in m.read().lower(): return True
+    except Exception: pass
+    return False
+
+def get_python_root() -> str:
```
Comment (zyunlam) [inline Test-Rack-Software/util/os_interface.py:13]: Comment to say that it only works for the raspberry pi

File: Test-Rack-Software/util/pio_commands.py [hunk 15]
```diff
@@ -17,38 +21,55 @@
 #### Helper functions ####
 
 
+def run_script_threaded(arg_list, callback: callable = None):
+    """Runs a platformio command in a subthread. Calls `callback` every iteration."""
+    thr = threading.Thread(target=run_script, args=(arg_list,))
+    thr.start()
+    while thr.is_alive():
+        time.sleep(0.02)
```
Comment (zyunlam) [inline Test-Rack-Software/util/pio_commands.py]: Comment why we need to sleep

**Review by zyunlam — CHANGES_REQUESTED** [review]

File: Test-Rack-Software/local.py
Comment (SuragNuthulapaty) [inline Test-Rack-Software/local.py:1]: why is this here?

File: Test-Rack-Software/util/dynamic_url.py (hunk 12 above)
Comment (SuragNuthulapaty) [reply Test-Rack-Software/util/dynamic_url.py:17]: why remove it? don't we want to remove any erroneous whitespace?

File: Test-Rack-Software/util/remote_command.py (hunk 8 above)
Comment (SuragNuthulapaty) [reply Test-Rack-Software/util/remote_command.py:38]: Probably no need, if we do or don't we are still going to have to download a new branch on each run. Bringing it back to master might just take slow down the whole process of recovering

File: Test-Rack-Software/util/pio_commands.py (hunk 15 above)
Comment (SuragNuthulapaty) [reply Test-Rack-Software/util/pio_commands.py]: why exactly do we need to sleep?

File: Test-Rack-Software/util/remote_command.py (hunk 8 above)
Comment (mpkarpov-ui) [reply Test-Rack-Software/util/remote_command.py:38]: Hmm.. I don't really have a counterargument to that other than "we're in an undefined remote state" after `reset`s, I'm good to keep it removed though. Thoughts @zyunlam ?

File: Test-Rack-Software/util/pio_commands.py (hunk 15 above)
Comment (mpkarpov-ui) [reply Test-Rack-Software/util/pio_commands.py]: I think I copied it over from some other threading code, we should explore if this works as intended without the sleep.

My guess is I put it there to not inundate the `callback` with calls so I limited it to only 50 hz, but we should look into it.

**Comment by mpkarpov-ui** [comment]
We've done pretty much all we can without validating that the MIDAS interface still works, so we should meet up to validate that our changes are good tomorrow before merging.

File: Test-Rack-Software/midas_rack/interface.py (hunk 4 above)
Comment (mpkarpov-ui) [reply Test-Rack-Software/midas_rack/interface.py:307]: We're going to keep this for now, it doesn't slow down the runs and lets us also buffer waiting for serial.

**Comment by SuragNuthulapaty** [comment]
LGTM!

**Review by SuragNuthulapaty — APPROVED** [review]
LGTM
